### ISSUE TITLE
ChatBedrock should support `default_headers`

### ISSUE THREAD

**Issue by efriis** [langchain-ai/langchain-aws#283]
similar to ChatAnthropic: https://python.langchain.com/api_reference/anthropic/chat_models/langchain_anthropic.chat_models.ChatAnthropic.html#langchain_anthropic.chat_models.ChatAnthropic.default_headers

**Comment by 3coins** [comment]
cc @guilt

**Comment by guilt** [comment]
@efriis Is the request to just have a parameter so that the callee does not error on an unknown parameter? (or) 
              to actually support the headers that Anthropic needs? 

For Bedrock Invoke, the headers actually would go into the body; For Converse, it would not go anywhere.

Could I request you to provide us sample usage code that we could test for supporting this behavior?

**Comment by efriis** [comment]
Was filing this on behalf of a community member - let me see if I can get them to respond with their needs! I think that behavior on invoke is quite confusing, so it might change what they're looking for.

**Comment by efriis** [comment]
I believe the ask is just to have the ability to attach custom headers to the request (potentially for tagging on an outbound proxy type thing) - is there not a concept of attaching headers in boto?

**Comment by guilt** [comment]
One of the good usecases for `ChatAnthropic` would be to [set up the version to invoke](https://docs.anthropic.com/en/api/versioning). That would no more be possible through the `boto` route, for instance - because those Anthropic headers move to the body in `InvokeModel`.

Yeah, we can add headers in `boto`. It needs more setup, because that header would need to be added, then signed. It has a particular lifecycle, [explained on the events guide here](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/events.html)

**Comment by mdfederici** [comment]
Yeah so for instance I want to be able to use prompt caching. https://docs.anthropic.com/en/docs/build-with-claude/prompt-caching. To turn on that beta feature you need to include a special header: 
![Screenshot 2024-12-12 at 8 07 36 AM](https://github.com/user-attachments/assets/7f3b64d1-bebe-4cca-bc17-ef37d5497e81). Ideally this would work through bedrock

**Comment by guilt** [comment]
@mdfederici Thank you for replying. Bedrock does support Prompt Caching and it works slightly differently [prompt-caching guide](https://docs.aws.amazon.com/bedrock/latest/userguide/prompt-caching.html). Is that the usecase you would like `BedrockChat` and `BedrockChatConverse` to get? Any other usecases? Please reply.

**Comment by mdfederici** [comment]
We frequently change models and providers with all of the parts of our app which are interacting with LLMs. So for instance I may have a piece of code which uses sonnet and one time I run from the UI I'll select sonnet on bedrock and another time i'll select sonnet on anthropic. We use langchain to take care of the routing (among many other things). Ideally we wouldn't have to change code or maintain a bunch of different ways to do this depending on if i am going to use bedrock, anthropic, openai, whatever. What I'd like from langchain is one consistent way to specify that I want prompt caching turned on regardless of whether i'm using chatbedrock or chatanthropic or whatever else. prompt caching is just an example of that, not a particularly important feature that i need to get enabled in my app.

**Comment by guilt** [comment]
Thank you for explaining this. How does the current usage of headers for Anthropic vs. OpenAI headers look like from your current usage point of view. Would like to understand that.

Currently though, will run this through @3coins to understand if there's a nice way to specify unified configs across various implementation providers if it ends up making **prompt caching** easy across different providers. And if there isn't a long term solution, we can for now work on enabling this in our two `Bedrock` implementations and extend it to a common interface when there will exist such an interface.

**Comment by afspies** [comment]
Hi, We've internally overriden ChatBedrock to support prompt caching by adding the new headers and changing some types etc. but I was curious to know if this is still something you guys are working on adding, or if we need to worry about long term stability?

**Comment by michaelnchin** [comment]
Implemented in ChatBedrockConverse with #811.
